core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java

```diff
@@ -1553,4 +1565,256 @@ public void setCacheProvider(CacheProvider cacheProvider) {
       reader.setCacheProvider(cacheProvider);
     }
   }
+
+  public static class RangedReader extends HeapIterator implements RFileReader {
```

Comment:
Is there anything forcing this to be inside RFile?  If it could be exist outside of RFile then it seems like it would be better for testing because RFile and this class could be tested independently. If it were separate then when Accumulo had ranges for an RFile then it could use this iterator on top of the RFile iterator.

Also want to understand if there is something forcing this design decision as that will help me review this further.

Reply:
I will need to look at it more when I have time later this week and test/verify to confirm but I think we should be able to move it outside of RFile. I mostly added it inside the RFile as that was where the other existing Reader was vs some actual need for it to be there. The biggest thing is FencedReader which is used inside of the RangedReader is package scope and any other protected things need to be accessible so if we move it outside the file we need to make sure those things are still visible.

Reply:
I looked around in the code to see where might be a good place to insert this iterator and I ended up creating #3259 as a result.  I think a new RangedIterator could be inserted using the new method added to DataFileValue using the ranges it has.  Looking around I also realized we are already setting up an optional per file iterator based on information from the metadata table.

Based on #3259 it seems we only need to worry about InterruptibleIterator and probably do not need to worry about implementing the more general FileSKVIterator interface.

Reply:
Sorry for the delayed response, was side tracked the last couple of weeks with some other things like #3286. #3259 definitely looks like an ideal place to hook into for creating the new ranged iterator as we will have access to the new ranges stored in DataFileValue that will be part of #3286. 

I can also go ahead and update this PR to only implement the InterruptibleIterator and not FileSKVIterator and see how it goes. I had initially used FileSKVIterator as that is what RFile used and I wanted to kind of make it a drop in replacement for a standard RFile and for tests but it doesn't really work to implement it after trying to do that. Most of the methods only make sense for the entire file and not a partial subset of a file, like get/first last key, so there's no good way to implement them. Also it doesn't seem like we need those methods when doing the actual scans/compactions anyways.